zenoh/src/api/bytes.rs

```diff
@@ -819,2538 +712,831 @@ impl std::fmt::Display for ZDeserializeError {
 }
-
-#[doc(hidden)]
-impl From<&ZSlice> for ZBytes {
-    fn from(t: &ZSlice) -> Self {
-        ZSerde.serialize(t)
+impl<'a, const N: usize> Serialize<'a> for [u8; N] {
+    type Error = Infallible;
+    fn write(self, writer: &mut ZBytesWriter) -> Result<(), Self::Error> {
+        std::io::Write::write(writer, &self).unwrap();
+        Ok(())
+    }
+    fn serialize(self) -> Result<Serialized<'a>, Self::Error> {
+        unimplemented!()
     }
 }
-
-#[doc(hidden)]
-impl Serialize<&mut ZSlice> for ZSerde {
-    type Output = ZBytes;
-
-    fn serialize(self, t: &mut ZSlice) -> Self::Output {
-        ZBytes::new(t.clone())
+impl<const N: usize> Deserialize<'_> for [u8; N] {
+    fn read(reader: &mut ZBytesReader) -> Result<Self, ZDeserializeError> {
+        let mut buf = [0; N];
+        std::io::Read::read_exact(reader, &mut buf).or(Err(ZDeserializeError))?;
+        Ok(buf)
+    }
+    fn deserialize(_zbytes: &'_ ZBytes) -> Result<Self, ZDeserializeError> {
+        unimplemented!()
+    }
+    fn try_from_zbytes(_zbytes: ZBytes) -> Result<Self, ZDeserializeError> {
+        unimplemented!()
     }
 }
-
-#[doc(hidden)]
-impl From<&mut ZSlice> for ZBytes {
-    fn from(t: &mut ZSlice) -> Self {
-        ZSerde.serialize(t)
+impl<const N: usize> From<[u8; N]> for ZBytes {
+    fn from(value: [u8; N]) -> Self {
+        ZBytes(value.into())
     }
 }
 
-#[doc(hidden)]
-impl Deserialize<ZSlice> for ZSerde {
-    type Input<'a> = &'a ZBytes;
+// Vec<u8>
+impl SerializedSize for Vec<u8> {}
+impl<'a> Serialize<'a> for Vec<u8> {
     type Error = Infallible;
-
-    fn deserialize(self, v: Self::Input<'_>) -> Result<ZSlice, Self::Error> {
-        Ok(v.0.to_zslice())
+    fn write(self, _writer: &mut ZBytesWriter) -> Result<(), Self::Error> {
+        unimplemented!()
     }
-}
-
-#[doc(hidden)]
-impl From<ZBytes> for ZSlice {
-    fn from(value: ZBytes) -> Self {
-        ZBuf::from(value).to_zslice()
+    fn serialize(self) -> Result<Serialized<'a>, Self::Error> {
+        Ok(ZBytes::from(self).into())
     }
 }
-
-#[doc(hidden)]
-impl From<&ZBytes> for ZSlice {
-    fn from(value: &ZBytes) -> Self {
-        ZSerde.deserialize(value).unwrap_infallible()
+impl<'a> Serialize<'a> for &'a Vec<u8> {
+    type Error = Infallible;
+    fn write(self, _writer: &mut ZBytesWriter) -> Result<(), Self::Error> {
+        unimplemented!()
     }
-}
-
-#[doc(hidden)]
-impl From<&mut ZBytes> for ZSlice {
-    fn from(value: &mut ZBytes) -> Self {
-        ZSerde.deserialize(&*value).unwrap_infallible()
+    fn serialize(self) -> Result<Serialized<'a>, Self::Error> {
+        Ok(self.as_slice().into())
     }
 }
-
-// [u8; N]
-impl<const N: usize> Serialize<[u8; N]> for ZSerde {
-    type Output = ZBytes;
-
-    fn serialize(self, t: [u8; N]) -> Self::Output {
-        ZBytes::new(t)
+impl Deserialize<'_> for Vec<u8> {
+    fn read(_reader: &mut ZBytesReader) -> Result<Self, ZDeserializeError> {
+        unimplemented!()
```

Comment:
It should be implemented.

Reply:
No it doesn't, because it should never be called, as the the semantic was:
- call `read` if the size is fixed (not the case for `Vec<T>`)
- call `deserialize`/`try_from_zbytes` if the size is variable